File: tree/diameteroftree.cpp
class Solution {
  public:
    int height(Node* root, int &ans){
    if(root == NULL){
        return 0;
    }
    int left_height= height(root->left,ans);// height of left subtree.
    int right_height= height(root->right,ans);//height of right subtree.
    
    ans= max(ans,1 + left_height + right_height);
    
    return max(left_height,right_height) + 1;    
    }
    //in this function we update the "ans" variable every time we get the height of left subtree and right subtree.and ofcoure returning the heigtht for the next subtree of node.  

    //the ans variable contains diameter of the tree.



    int diameter(Node* root) {
        
        if(root==NULL){
            return 0;
        }
        
        int ans;
        height(root,ans);
        
        return ans;
    }
    // in this we will call the height function and we will get the diameter of tree
    //so we have to just return the ans variable which contains the diameter of the tree.
};

File: tree/diameteroftree.js
var diameterOfBinaryTree = function(root) {
     if(!root){
        return 0;
     }
     let maxDiameter =0;

     const dfs =(node)=>{
        if(!node){
            return 0;
        }
        let left = dfs(node.left);
        let right =dfs(node.right);

        maxDiameter= Math.max(maxDiameter,left+right);

        return Math.max(left,right) + 1;
     }
     dfs(root);

     return maxDiameter;

};

